programs/token-2022/src/extension/transfer_hook/state.rs

```diff
@@ -0,0 +1,155 @@
+use core::mem::MaybeUninit;
+use crate::{write_bytes, UNINIT_BYTE, ID};
+use crate::extension::consts::ExtensionDiscriminator;
+use pinocchio::{
+    account_info::{AccountInfo, Ref},
+    program_error::ProgramError,
+    pubkey::Pubkey,
+};
+
+#[repr(u8)]
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub enum TransferHookInstruction {
+    Initialize,
+    Update,
+}
+
+#[repr(C)]
+#[derive(Clone, Copy, Debug, Default, PartialEq)]
+pub struct TransferHook {
+    /// Authority that can set the transfer hook program id
+    authority: Pubkey,
+    /// Program that authorizes the transfer
+    program_id: Pubkey,
+}
+
+impl TransferHook {
+    pub const AUTHORITY_START: usize = 170;
+
+    pub const BASE_LEN: usize = core::mem::size_of::<TransferHook>();
+
+    pub const LEN: usize = Self::AUTHORITY_START + Self::BASE_LEN;
+
+    /// Return a `TransferHook` from the given bytes (unsafe, unchecked).
+    #[inline(always)]
+    pub unsafe fn from_bytes_unchecked(bytes: &[u8]) -> &Self {
+        &*(bytes[Self::AUTHORITY_START as usize..].as_ptr() as *const TransferHook)
```

Comment:
It is a bit strange to use a fix offset to where the extension will start since it depends on what other extensions are enabled. This is why the original token-2022 code looks up where the extension data is: https://github.com/solana-program/token-2022/blob/main/interface/src/extension/mod.rs#L148-L183